cpp/dolfinx/refinement/interval.h

```diff
@@ -0,0 +1,235 @@
+  std::vector<std::int64_t> global_indices
+      = adjust_indices(*mesh.topology()->index_map(0), number_of_refined_edges);
+
+  // Build the topology on the new vertices
+  const auto refined_cell_count = mesh.topology()->index_map(1)->size_local()
+                                  + mesh.topology()->index_map(1)->num_ghosts()
+                                  + number_of_refined_edges;
+
+  std::vector<std::int64_t> edge_topology;
+  edge_topology.reserve(refined_cell_count * 2);
+
+  std::vector<std::int32_t> parent_edge;
+  parent_edge.reserve(refined_cell_count);
+
+  for (std::int32_t edge = 0; edge < map_e->size_local(); ++edge)
+  {
+    const auto& vertices = e_to_v->links(edge);
+    assert(vertices.size() == 2);
+
+    // we consider a (previous) edge of (global) vertices
+    // a ----------- b
+    const std::int64_t a = global_indices[vertices[0]];
+    const std::int64_t b = global_indices[vertices[1]];
+
+    if (refinement_marker[edge])
+    {
+      // find (global) index of new midpoint vertex:
+      // a --- c --- b
+      auto it = new_vertex_map.find(edge);
+      assert(it != new_vertex_map.end());
+      const std::int64_t c = it->second;
+
+      // add new edges to refined topology
+      edge_topology.insert(edge_topology.end(), {a, c, c, b});
+      parent_edge.insert(parent_edge.end(), {edge, edge});
+    }
+    else
+    {
+      // copy the previous edge
+      edge_topology.insert(edge_topology.end(), {a, b});
+      parent_edge.push_back(edge);
+    }
+  }
+
+  assert(edge_topology.size() == refined_cell_count * 2);
+  assert(parent_edge.size() == refined_cell_count);
+
+  std::vector<std::int32_t> offsets(refined_cell_count + 1);
+  std::ranges::generate(offsets, [i = 0]() mutable { return 2 * i++; });
+
+  graph::AdjacencyList cell_adj(std::move(edge_topology), std::move(offsets));
+
+  return {std::move(cell_adj), std::move(new_vertex_coords), xshape,
+          std::move(parent_edge)};
+}
+
+} // namespace impl
+
+/// Refines a (topologically) one dimensional mesh by splitting edges.
+///
+/// @param[in] mesh Mesh to be refined
+/// @param[in] edges Optional indices of the edges that should be split by this
+/// refinement. If not provided, all edges are considered marked for refinement,
+/// i.e. a uniform refinement is performed.
+/// @param[in] redistribute Option to enable redistribution of the refined mesh
+/// across processes.
+///
+/// @return Refined mesh, and list of parent edges - for every new edge index
+/// this contains the associated edge index of the pre-refinement mesh.
+template <std::floating_point T>
+std::tuple<mesh::Mesh<T>, std::vector<std::int32_t>>
+refine_interval(const mesh::Mesh<T>& mesh,
+                std::optional<std::span<const std::int32_t>> edges,
+                bool redistribute)
+{
+
+  if (mesh.topology()->cell_type() != mesh::CellType::interval)
+    throw std::runtime_error("Cell type not supported");
+
+  if (!mesh.topology()->index_map(1))
```

Comment:
I think this can just be an `assert` as it must exist, unless there is an internal error.